### ISSUE TITLE
Using a pen and keyboard on mobile does not remove Touchscreen nerf. Why?

### ISSUE THREAD

**Issue by dokkkku** [ppy/osu#27309]
### Type

Game behaviour

### Bug description

Title and Video explain pretty much everything already but I'll write anyway.

Basically, i was using my s pen and my keypad on my samsung tablet on Lazer, and i disabled touchscreen inputs and used my pen to drag and keypad to click the circles. This playstyle does not have any advantages over regular playstyle (i actually find it really hard even tho i used to play like this), yet it still gets detected as touchscreen and the pp/SR nerf gets applied which i find really stupid.

### Screenshots or videos

https://github.com/ppy/osu/assets/66845810/9de29300-e698-4767-8f4d-48f856935b30

### Version

2024.221.0-lazer

### Logs

[logs.zip](https://github.com/ppy/osu/files/14370621/logs.zip)

**Comment by bdach** [comment]
lazer is using a different detection method to stable. stable's was based solely off of the inputs the player made. lazer's based on the actual backing hardware. Either the pen isn't reporting that it's a pen rather than a touch, or we're not handling that correctly. I can't tell because I don't own any of the hardware.

**Comment by dokkkku** [comment]
> lazer is using a different detection method to stable. stable's was based solely off of the inputs the player made. lazer's based on the actual backing hardware. Either the pen isn't reporting that it's a pen rather than a touch, or we're not handling that correctly. I can't tell because I don't own any of the hardware.

I'm aware how Lazer detects inputs differently from stable. Though I can't it confusing as to why it would determine the nerf on sorely the hardware used and not the way the map was played. I'm in no way a programmer, but i would think that a system for detecting whether the cursor is constantly teleporting or not could be implemented (i would also assume that's how stanle did it)

**Comment by bdach** [comment]
Yes we could port stable's algorithm. The main issue with that is that stable's algorithm is basically a heuristic. Hardware-based detection is preferred if it can be made to work.

**Comment by eclecticfluff** [comment]
There are already external drawing tablets with screens. If you wabted to ban such devices, I doubt there would be a way to accurately judged when one was used. Also, even if being able to mentally bridge the gap between input device movenent and cursor movenent is a significant diffuculty factor (which I doubt), the fact that this movement must be over a greater area or else run into issues with not being able to see coming notes nullifies the advantage in any map with apreciable difficulty.

As for coming up with a heuristic so that even a passive stylus could be used on a tablet computer and not get a lower multiplier, would it not be possible to keep track of touch events and only declare a play as touchscreen when the touch for guiding the cursor is released or changes? I am not a software developer so am likely missing things, but this sounds simple and relibale enough to impliment without too much issue from my outsider′s perspective.

**Comment by eclecticfluff** [comment]
As for apple pencil, doing a quick search of the documentation, while its inputs are reported as standard tluch events, there is some additional information about angle that also gets transmitted which could be used for hardware detection.

https://developer.apple.com/documentation/uikit/pencil_interactions/handling_input_from_apple_pencil

**Comment by Rekunan** [comment]
> If you wanted to ban such devices

I don't think they should be banned, but they should be receiving the TD mod. For supporting evidence:

> (which I doubt)

I think the best evidence is basically every digital artist preferring screen tablets over pen tablets, but I also have my own anecdotal experience from the arcade, where I always perform better on cabinets with the mental mapping requirement removed, such as chunithm, maimai, and jubeat, compared to ones without it removed, such as popn, sdvx, and iidx.

I have actually encountered the hand blocking screen issue on jubeat and also on maimai screen notes. You are right in that it has at least some nullification, however, it was surprisingly easy to overcome that and start performing way better.

**Comment by shiumano** [comment]
The most important matter to check personally is what is the excellence of the TD.
The most important element of the TD is that the input device and video output are identical, which is also true for tablets with LCD displays.
And second, this is the most important issue. That is that there are multiple cursors.
If you want to go back and forth between the right and left edges of the playfield 5 times per second, there is nothing wrong if you use your right and left hands.
Thus, Aim (especially jumping) in TD deserves to be underestimated.
However, touch input in osu!lazer is more than just hitting the notes directly.
You can play with one hand on the stylus and the other hand on the keys, in other words, no different from a tablet with a screen.
The switching between the two is done programmatically and can be easily determined.
One suggestion: "No TDs are attached at the start of play. How about a flow of "no TD is put on at the start of play, but once Direct Touch occurs, TD is put on without question"?
This would require the ability to add mods during gameplay, which would also solve other edge cases. (e.g. "start with keyboard and mouse, play with touch screen, but no TD Mod is attached" or "clear with TD, retry with keyboard and mouse, but TD is still attached")
I know this is a lot to say and using the DeepL translation, so I know it is not always easy to understand, but please consider it.

**Comment by P4blohh** [comment]
Just thought i would chime in on this, since it seems that a large portion of people are unaware of just how much TDP (Touch Device Pen) differs from more conventional playstyles such as mouse and tablet. Here is a short text i wrote regarding a player named Bonk, who utilizes TDP (although he uses his finger instead of a pen):  
"Mechanically it might seem that his playstyle functions the same way as other conventional playstyles (mouse and tablet), but fundamentally his playstyle bypasses an entire portion of the reading process that is unique to those playstyles, and for that reason alone it makes no sense for him to be on the same leaderboard as mouse and tablet players. It cannot be understated how crucial the degree of separation between your cursor and input device is, as it requires you to translate your reading into tablet / mouse movements that THEN get translated into cursor movements. This is both a mental and physical process that requires a HUGE deal of muscle memory to perfect. Bonks playstyle skips that process almost entirely. His input device allows him to aim the notes directly, as his hand movements don’t have to go through a medium before getting translated into cursor movements. In other words, his input device movements ARE his cursor movements. Not only does this undermine the perceived difficulty of many of the aim control maps he's playing, but it's also just an entirely different playstyle that differs heavily from mouse and tablet. In conclusion, they aren't legit if you're comparing them to plays set on mouse and tablet"
Now obviously, just like with any other playstyle, TDP does have it's own downsides, and while some of those downsides are pretty challenging to deal with, i don't believe they come anywhere near close to offsetting the one monumental change that comes with TDP. I personally believe players using TDP should have their own separate leaderboards, as i find it nonsensical to have fundamentally different playstyles on the same leaderboard.

**Comment by trustytrojan** [comment]
> I personally believe players using TDP should have their own separate leaderboards

The osu! team is probably aware of this, hence the existence of the TD mod...
Filter leaderboards by TD and you're set. Sure max score isn't 1 million but it's not hard to get adjusted to.
I do hope leaderboard filtering by mod becomes a **free** feature in the future (global only), as it's starting to seem more essential to lazer as a whole with the amount of mods it's accrued over its lifespan.
